src/libs/elektra/changetracking.c

```diff
@@ -0,0 +1,601 @@
+	KDB * kdb = kdbPtr == NULL ? NULL : *(KDB **) keyValue (kdbKey);
+
+	return elektraChangeTrackingGetContextFromKdb (kdb);
+}
+
+/**
+ * @internal
+ *
+ * @brief For testing purposes only: Create a ChangeTrackingContext from a keyset
+ *
+ * @param oldKeys the old keys
+ * @return a new changetracking context
+ */
+ChangeTrackingContext * elektraChangeTrackingCreateContextForTesting (KeySet * oldKeys)
+{
+	ChangeTrackingContext * context = elektraCalloc (sizeof (ChangeTrackingContext));
+
+	ksIncRef (oldKeys);
+	context->oldKeys = oldKeys;
+
+	return context;
+}
+
+/**
+ * @internal
+ *
+ * @brief For testing purposes only: Delete a ChangeTrackingContext object
+ *
+ * @param context the object to delete
+ */
+void elektraChangeTrackingContextDel (ChangeTrackingContext * context)
+{
+	if (context->oldKeys != NULL)
+	{
+		ksDecRef (context->oldKeys);
+		ksDel (context->oldKeys);
+		context->oldKeys = NULL;
+	}
+
+	elektraFree (context);
+}
+
+/**
+ * Determines whether two keys have different value
+ *
+ * @param new new key
+ * @param old old key
+ * @return true if @p new and @p old have a different value
+ */
+static bool keyValueDifferent (Key * new, Key * old)
+{
+	if (new->keyData == old->keyData)
+	{
+		return false;
+	}
+
+	if (new->keyData == NULL || old->keyData == NULL)
+	{
+		// one is NULL, the other one isn't -> modified
+		return true;
+	}
+
+	if (keyIsString (new) != keyIsString (old))
+	{
+		// different data types -> modified
+		return true;
+	}
+
+	if (keyIsString (new))
+	{
+		if (strcmp (keyString (new), keyString (old)) != 0)
+		{
+			// different value -> modified
+			return true;
+		}
+	}
+	else
+	{
+		if (new->keyData->dataSize != old->keyData->dataSize)
+		{
+			// different datasize -> modified
+			return true;
+		}
+
+		if (memcmp (new->keyData->data.v, old->keyData->data.v, new->keyData->dataSize) != 0)
+		{
+			// different value -> modified
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/**
+ * Find the differences between two keysets
+ *
+ * @param new the new keyset
+ * @param old the old keyset
+ * @param addedKeys adds keys present in @p new but not in @p old
+ * @param removedKeys adds keys present in @p old but not in @p new
+ * @param modifiedKeys adds keys present in both @p new and @p old, but with changes in value or in the meta keys
+ * @param parentKey parent key - if this parameter is not NULL, only keys below or same are processed.
+ */
+static void findDifferences (KeySet * new, KeySet * old, KeySet * addedKeys, KeySet * removedKeys, KeySet * modifiedKeys,
+			     const Key * parentKey)
+{
+	KeySet * metaAdded = ksNew (0, KS_END);
+	KeySet * metaRemoved = ksNew (0, KS_END);
+	KeySet * metaModified = ksNew (0, KS_END);
+
+	for (elektraCursor itOld = 0; itOld < ksGetSize (old); itOld++)
+	{
+		Key * needle = ksAtCursor (old, itOld);
+
+		if (parentKey != NULL && !keyIsBelowOrSame (parentKey, needle))
+		{
+			continue;
+		}
+
+		Key * found = ksLookup (new, needle, 0);
```

Comment:
Of course I don't have benchmarks to proof it, but this `ksLookup` looks very much like a bottleneck. Both `new` and `old` are sorted the same way. You could just iterate them both at the same time, and compare the keys with `keyCmp (oldKey, newKey)`:
- If both have the same name (`keyCmp == 0`), do a comparison (`else` branch below). Increment both cursors.
- If `keyCmp(oldKey, newKey) < 0`, `oldKey` was removed from `new`. Increment just cursor for `old` and check again.
- If `keyCmp(oldKey, newKey) > 0`, `newKey` was added to `new`. Increment just cursor for `new` and check again.

Reply:
Good catch! It took me a while to implement, but you're right. Based on some benchmarking (500 keys - 100 unmodified, 100 modified, 100 added, 100 removed) the current implementation averages at about ~350 microseconds, and your proposed implementation at about 27 microseconds.